mttl/config.py

```diff
@@ -1,89 +1,140 @@
+                new_fields[f.name] = value
+            else:
+                value = new_fields[f.name][1]
+                value.default.update(c, f.default, f.type)
 
-        if kwargs:
-            overwrite_logs += self.update_kwargs(
-                kwargs, raise_error=raise_error, silent=silent
-            )
+    # now we need to set the default values for all the fields in which there is only one default value
+    for k in new_fields:
+        value = new_fields[k][1]
+        value.default = value.default.default
 
-        # setup logging to the output dir
-        try:
-            setup_logging(self.output_dir)
-        except Exception as e:
-            if raise_error:
-                raise ValueError("Error setting up logging") from e
-            elif not silent:
-                logger.warning(f"Error setting up logging to {self.output_dir}")
+    return [(k,) + v for k, v in new_fields.items()]
 
-        # log the overwrites
-        for log in overwrite_logs:
-            logger.warning(log)
 
-        self.post_init(silent=silent)
+def create_config_class_from_args(config_class: dataclass, args: "Args"):
+    """
+    Load a dataclass from the arguments.
+    """
+    kwargs = {}
 
-    def post_init(self, silent=False):
-        if self.attn_implementation == "eager" and self.pack_sequences:
-            logger.warning(
-                "Eager attention is not compatible with packed sequences"
-                + ", tokens across examples will not be masked"
-            )
-        elif self.attn_implementation == "flash_attention_2" and self.pack_sequences:
-            logger.warning(
-                "The wrapper we provide for flash attention 2 may not behave as expected for"
-                + " some models. Please make sure you test the model with packed sequences"
-            )
+    for f in fields(config_class):
+        if hasattr(args, f.name):
+            value = getattr(args, f.name)
+
+            # if the field is not a multi-default, we override it (it was either default or set by the user)
+            if not isinstance(value, MultiDefaultValue):
+                kwargs[f.name] = value
+
+    return config_class(**kwargs)
+
+
+@dataclass
+class Args:
+    @property
+    def updated_kwargs(self):
+        return {
+            k.name: getattr(self, k.name)
+            for k in fields(self)
+            if getattr(self, k.name) != k.default
+        }
 
     @classmethod
-    def fromdict(cls, data):
-        _ = data.pop("_updated_kwargs", None)
-        return cls(kwargs=data, raise_error=False, silent=True)
+    def fromdict(cls, data, strict=False) -> "Args":
+        """
+        Reload a dataclass from a dictionary. We store the class name in the dict to be able to reload it.
+        If the cls is Args, then we try to access the `_class` attribute to get the class name.
+        """
+        cls_name = data.pop("args_class", None)
+        if cls == Args:
+            if not cls_name:
```

Comment:
when would you do `Args.fromdict` instead of `<subclass>.from_dict` ?

Reply:
when you reload from the raw dictionary, e.g. maybe some experts have been trained with MoETrainingConfig and you want to inspect some of those argument values